fendermint/testing/materializer/src/docker/node.rs

```diff
@@ -0,0 +1,566 @@
+                        node_config.network,
+                        vec![
+                            (port_range.cometbft_p2p_host_port(), COMETBFT_P2P_PORT),
+                            (port_range.cometbft_rpc_host_port(), COMETBFT_RPC_PORT),
+                        ],
+                        entrypoint("cometbft start"),
+                    )
+                    .await
+                    .context("failed to create fendermint")?
+            }
+        };
+
+        // Create a ethapi container
+        let ethapi = match ethapi {
+            None if node_config.ethapi => {
+                let creator =
+                    DockerRunner::new(&dh, node_name, FENDERMINT_IMAGE, user, volumes(vec![]));
+
+                let c = creator
+                    .create(
+                        ethapi_name,
+                        node_config.network,
+                        vec![(port_range.ethapi_rpc_host_port(), ETHAPI_RPC_PORT)],
+                        entrypoint("fendermint eth run"),
+                    )
+                    .await
+                    .context("failed to create ethapi")?;
+
+                Some(c)
+            }
+            other => other,
+        };
+
+        // Construct the DockerNode
+        Ok(DockerNode {
+            node_name: node_name.clone(),
+            fendermint,
+            cometbft,
+            ethapi,
+            port_range,
+            path: node_dir,
+        })
+    }
+}
+
+/// Create a container name from a node name and a logical container name, e.g. "cometbft"
+/// in a way that we can use it as a hostname without being too long.
+///
+/// It consists of `{node-id}-{container}-{hash(node-name)}`,
+/// e.g. "node-12-cometbft-a1b2c3"
+fn container_name(node_name: &NodeName, container: &str) -> String {
+    let node_id = node_name
+        .path()
+        .file_stem()
+        .unwrap_or_default()
+        .to_string_lossy()
+        .to_string();
+    let hash = ResourceHash::digest(node_name.path().to_string_lossy().to_string());
+    let hash = hash.to_string();
+    let hash = &hash.as_str()[..6];
+    format!("{node_id}-{container}-{}", hash)
+}
+
+struct DockerRunner<'a> {
+    dh: &'a DockerWithDropHandle,
+    node_name: NodeName,
+    image: String,
+    user: u32,
+    volumes: Volumes,
+}
+
+impl<'a> DockerRunner<'a> {
+    pub fn new(
+        dh: &'a DockerWithDropHandle,
+        node_name: &NodeName,
+        image: &str,
+        user: u32,
+        volumes: Volumes,
+    ) -> Self {
+        Self {
+            dh,
+            node_name: node_name.clone(),
+            image: image.to_string(),
+            user,
+            volumes,
+        }
+    }
+
+    fn docker(&self) -> &Docker {
+        &self.dh.docker
+    }
+
+    // Tag containers with resource names.
+    fn labels(&self) -> HashMap<String, String> {
+        [
+            ("testnet", self.node_name.testnet().path()),
+            ("node", self.node_name.path()),
```

Comment:
would it make sense to also have a `container_name` label?

Reply:
Depends on what exactly that is. You mean something to tell which of the three it is, like "I want to find all ethapi containers"? You don't mean the unique name produced by the `container_name()` function, right? Because that will be the actual name, so there is no reason to add it as metadata.